File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        
        if (!head) {
            return nullptr;
        }
        
        ListNode* curr = head;

        vector<ListNode*> all_nodes = {};

        while (curr != nullptr) {
            cout << "The current node value is: " << curr->val <<endl;
            all_nodes.push_back(curr);
            curr = curr->next;
        }

        // Now curr is pointing to the last element
        for (int i = all_nodes.size() - 1; i > 0; i--) {
            cout << "The nodes are: " << all_nodes[i]->val <<endl;
            all_nodes[i]->next = all_nodes[i-1];
        }

        // Making the 'new' last node next to nullptr
        all_nodes[0]->next = nullptr;

        return all_nodes[all_nodes.size()-1];
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        
        #finding the length of the list
        temp = head
        length = 0
        while (temp):
            temp = temp.next
            length += 1

        print(length)

        curr = head
        prev = None

        for i in range(length - n):
            if (curr.next):
                prev = curr 
                curr = curr.next
        
        #two cases, can be the end of the list or can be the middle
        if (curr.next):
            if (prev):
                prev.next = curr.next
                curr = None
            else:
                head = curr.next
        else:
            if (prev):
                prev.next = None
                curr = None
            else:
                return None
        
        return head
        